backend/eventloop/application_event_loop/application_event_loop.go

```diff
@@ -131,240 +157,251 @@ func applicationEventQueueLoop(ctx context.Context, k8sClient client.Client,
 	defer log.Info("applicationEventQueueLoop ended.")
 
-	// Only one deployment event is processed at a time
-	// These are events that are waiting to be sent to application_event_runner_deployments runner
-	var activeDeploymentEvent *RequestMessage
-	waitingDeploymentEvents := []*RequestMessage{}
-
-	// Only one sync operation event is processed at a time
-	// For example: if the user created multiple GitOpsDeploymentSyncRun CRs, they will be processed to completion, one at a time.
-	// These are events that are waiting to be sent to application_event_runner_syncruns runner
-	var activeSyncOperationEvent *RequestMessage
-	waitingSyncOperationEvents := []*RequestMessage{}
+	state := applicationEventQueueLoopState{
+		activeDeploymentEvent:      nil,
+		waitingDeploymentEvents:    []*RequestMessage{},
+		activeSyncOperationEvent:   nil,
+		waitingSyncOperationEvents: []*RequestMessage{},
 
-	deploymentEventRunner := aerFactory.createNewApplicationEventLoopRunner(input, sharedResourceEventLoop, gitopsDeploymentName,
-		gitopsDeploymentNamespace, workspaceID, "deployment")
-	deploymentEventRunnerShutdown := false
+		deploymentEventRunner: aerFactory.createNewApplicationEventLoopRunner(input, sharedResourceEventLoop, gitopsDeploymentName,
+			gitopsDeploymentNamespace, workspaceID, "deployment"),
+		deploymentEventRunnerShutdown: false,
 
-	syncOperationEventRunner := aerFactory.createNewApplicationEventLoopRunner(input, sharedResourceEventLoop, gitopsDeploymentName,
-		gitopsDeploymentNamespace, workspaceID, "sync-operation")
-	syncOperationEventRunnerShutdown := false
+		syncOperationEventRunner: aerFactory.createNewApplicationEventLoopRunner(input, sharedResourceEventLoop, gitopsDeploymentName,
+			gitopsDeploymentNamespace, workspaceID, "sync-operation"),
+		syncOperationEventRunnerShutdown: false,
+	}
 
 	// Start the ticker, which will -- every X seconds -- instruct the GitOpsDeployment CR fields to update
 	startNewStatusUpdateTimer(ctx, k8sClient, input, log)
 
-out:
 	for {
 		// Block on waiting for more events for this application
 		newEvent := <-input
 
-		eventLoopMessage := newEvent.Message.Event
+		// The event loop will be terminated if the application no longer exists
+		if terminateEventLoop := processApplicationEventQueueLoopMessage(ctx, newEvent, &state, input, k8sClient, log); terminateEventLoop {
+			break
+		}
+	}
+}
 
-		if eventLoopMessage != nil {
-			if !(eventLoopMessage.EventType == eventlooptypes.UpdateDeploymentStatusTick && disableDeploymentStatusTickLogging) {
-				log.V(logutil.LogLevel_Debug).Info("applicationEventQueueLoop received event",
-					"event", eventlooptypes.StringEventLoopEvent(eventLoopMessage))
+// returns true if the event loop should be terminated, false otherwise.
+func processApplicationEventQueueLoopMessage(ctx context.Context, newEvent RequestMessage, state *applicationEventQueueLoopState, input chan RequestMessage, k8sClient client.Client, log logr.Logger) bool {
```

Comment:
Note for reviewer: I moved the contents of `applicationEventQueueLoop` into  its own function, so it was easier to unit test. The code is otherwise the same (the GitHub compare is mangling it :smile: )